### PR TITLE
[Vertical compaction] Support primary key table

### PR DESCRIPTION
**Pull request by wyb**
#1112 

Primary key columns will be merged and written into the rowset segments first.
Then the non-key columns are merged by groups.

### CODE REVIEW

**Comment by wyb** [comment]
run starrocks_be_unittest

**Review by decster — APPROVED** [review]

File: be/src/storage/vectorized/rowset_merger.cpp [hunk 1]
```diff
@@ -222,25 +236,65 @@ class RowsetMergerImpl : public RowsetMerger {
...
+            RETURN_IF_ERROR(_do_merge_horizontally(tablet, version, schema, rowsets, writer, cfg, &total_input_size,
+                                                   &total_rows, &total_chunk, &stats));
+        }
+        timer.stop();
+
+        StarRocksMetrics::instance()->update_compaction_deltas_total.increment(rowsets.size());
+        StarRocksMetrics::instance()->update_compaction_bytes_total.increment(total_input_size);
+        StarRocksMetrics::instance()->update_compaction_outputs_total.increment(1);
+        StarRocksMetrics::instance()->update_compaction_outputs_bytes_total.increment(writer->total_data_size());
+        LOG(INFO) << "compaction merge finished. tablet=" << tablet.tablet_id() << " #key=" << schema.num_key_fields()
```
[the remark this answers is not in the document]
Comment (decster) [reply be/src/storage/vectorized/rowset_merger.cpp:264]: This is just merge finished, compaction has not finished

File: be/src/storage/vectorized/rowset_merger.cpp [hunk 2]
```diff
@@ -266,63 +321,195 @@ class RowsetMergerImpl : public RowsetMerger {
...
+    Status _do_merge_vertically(Tablet& tablet, int64_t version, const vector<RowsetSharedPtr>& rowsets,
+                                RowsetWriter* writer, const MergeConfig& cfg,
+                                const vector<vector<uint32_t>>& column_groups, size_t* total_input_size,
+                                size_t* total_rows, size_t* total_chunk, OlapReaderStatistics* stats) {
+        DCHECK_GT(column_groups.size(), 1);
+        // merge key columns
+        auto mask_buffer = std::make_unique<RowSourceMaskBuffer>(tablet.tablet_id(), tablet.data_dir()->path());
+        {
+            Schema schema = ChunkHelper::convert_schema_to_format_v2(tablet.tablet_schema(), column_groups[0]);
+            RETURN_IF_ERROR(_do_merge_horizontally(tablet, version, schema, rowsets, writer, cfg, total_input_size,
```
[the remark this answers is not in the document]
Comment (wyb) [reply be/src/storage/vectorized/rowset_merger.cpp:415]: Generate row source masks when merging key columns.

File: be/src/storage/vectorized/rowset_merger.cpp [hunk 3]
```diff
@@ -266,63 +321,195 @@ class RowsetMergerImpl : public RowsetMerger {
...
+            vector<vectorized::ChunkIteratorPtr> iterators;
+            iterators.reserve(rowsets.size());
+            OlapReaderStatistics non_key_stats;
+            Schema schema = ChunkHelper::convert_schema_to_format_v2(tablet.tablet_schema(), column_groups[i]);
+            for (const auto& rowset : rowsets) {
+                _entries.emplace_back(new MergeEntry<T>());
+                MergeEntry<T>& entry = *_entries.back();
+                entry.rowset_release_guard = std::make_unique<RowsetReleaseGuard>(rowset);
+                auto beta_rowset = down_cast<BetaRowset*>(rowset.get());
+                auto res = beta_rowset->get_segment_iterators2(schema, tablet.data_dir()->get_meta(), version,
```
[the remark this answers is not in the document]
Comment (wyb) [reply be/src/storage/vectorized/rowset_merger.cpp:436]: Already checked below.

**Review by chaoyli — APPROVED** [review]

**Pull request merged**